cellmlmanip/singularity_fixes.py

```diff
@@ -0,0 +1,219 @@
+
+    if not expr.has(exp_function):  # Expressions without exp don't have GHK equations
+        return (None, None, None, expr, False)
+
+    elif isinstance(expr, Add):  # A + B + ..
+        # The expression is an addition, find singularities in each argument
+        new_expr_parts = []
+        for a in expr.args:
+            new_expr_parts.append(_new_expr_parts(a, V, U_offset, exp_function))
+
+        # If all arguments have the same singularity point, return 1 singularity with the widest range
+        range = [item for (vs, ve, _, _, _) in new_expr_parts for item in (vs, ve)]
+        if len(new_expr_parts) > 1 and len(set([str(sp) for (_, _, sp, _, _) in new_expr_parts])) == 1 \
+                and all(isinstance(b, Quantity) for b in range):
+            sp = new_expr_parts[0][2]
+            vs, ve = min(range, key=lambda v: float(str(v))), max(range, key=lambda v: float(str(v)))
+            return (vs, ve, sp, expr, True)
+        else:  # Singularity points differ, so create each piecewise seperately and add them up
+            expr_parts = []
+            is_piecewise = False
+            for vs, ve, sp, ex, has_piecewise in new_expr_parts:
+                is_piecewise = is_piecewise or has_piecewise or vs is not None
+                expr_parts.append(_generate_piecewise(vs, ve, sp, ex, V))
+            return (None, None, None, Add(*expr_parts), is_piecewise)
+
+    elif isinstance(expr, Pow) and expr.args[1] == -1.0 and len(expr.args) == 2:  # 1/A
+        # Find singularities in A and adjust result to represent 1 / A
+        vs, ve, sp, ex, has_piecewise = _new_expr_parts(expr.args[0], V, U_offset, exp_function)
+        has_piecewise = has_piecewise or vs is not None
+        return (None, None, None, ONE / _generate_piecewise(vs, ve, sp, ex, V), has_piecewise)
+
+    elif isinstance(expr, Mul):  # A * B * ...
+        (vs, ve, sp) = _get_U(check_U_expr, V, U_offset, exp_function)  # Find the singularity point
+        if vs is not None:
+            return (vs, ve, sp, expr, True)
+        else:  # Couldn't find singularity, try the expression's arguments
+            expr_parts = []
+            is_piecewise = False
+            for sub_ex in expr.args:
+                vs, ve, sp, ex, has_piecewise = _new_expr_parts(sub_ex, V, U_offset, exp_function)
+                has_piecewise = is_piecewise = is_piecewise or has_piecewise or vs is not None
+                expr_parts.append(_generate_piecewise(vs, ve, sp, ex, V))
+            return (None, None, None, Mul(*expr_parts), is_piecewise)
+
+    else:  # Different type of equation e.g a number
+        return (None, None, None, expr, False)
+
+
+def new_expr(expr, V, U_offset=1e-7, exp_function=exp):
+    """Removes suitable singularities and replaces it with a piecewise.
+    :param: expr the expression to analyse ().
+    :param: V the volatge variable
+    :param: U_offset determins the offset either side of U for which the fix is used
```

Comment:
What's U?

Reply:
added reference to other function that explains it